File: src/lib/learning/vector_index/local_embed.py
#!/usr/bin/env python3
"""
Local Embeddings Generator

Generates 384-d embeddings using sentence-transformers.
Model: all-MiniLM-L6-v2

Usage:
    python local_embed.py --text="example text" --output=embedding.json
    python local_embed.py --batch_file=texts.jsonl --output=embeddings.jsonl

Input (batch mode):
    {"id": "001", "text": "example"}
    {"id": "002", "text": "another example"}

Output:
    {"id": "001", "embedding": [0.123, -0.456, ...]}
"""

import argparse
import json
import sys
from typing import List, Dict, Any

try:
    from sentence_transformers import SentenceTransformer
except ImportError:
    print("ERROR: sentence-transformers not installed. Run: pip install sentence-transformers", file=sys.stderr)
    sys.exit(1)

# Default model
DEFAULT_MODEL = "sentence-transformers/all-MiniLM-L6-v2"
EXPECTED_DIM = 384


def load_model(model_name: str) -> SentenceTransformer:
    """Load sentence-transformers model."""
    print(f"[Local Embed] Loading model: {model_name}", file=sys.stderr)
    model = SentenceTransformer(model_name)
    print(f"[Local Embed] Model loaded (dim={EXPECTED_DIM})", file=sys.stderr)
    return model


def embed_single(text: str, model: SentenceTransformer) -> List[float]:
    """Embed a single text."""
    embedding = model.encode(text, convert_to_tensor=False, show_progress_bar=False)
    return embedding.tolist()


def embed_batch(items: List[Dict[str, Any]], model: SentenceTransformer) -> List[Dict[str, Any]]:
    """Embed a batch of texts."""
    texts = [item["text"] for item in items]
    embeddings = model.encode(texts, convert_to_tensor=False, show_progress_bar=True, batch_size=32)
    
    results = []
    for idx, item in enumerate(items):
        results.append({
            "id": item["id"],
            "embedding": embeddings[idx].tolist(),
        })
    
    return results


def main():
    parser = argparse.ArgumentParser(description="Generate local embeddings")
    parser.add_argument("--text", help="Single text to embed")
    parser.add_argument("--batch_file", help="JSONL file with batch items")
    parser.add_argument("--output", help="Output file (JSON or JSONL)")
    parser.add_argument("--model", default=DEFAULT_MODEL, help="Model name")
    
    args = parser.parse_args()
    
    if not args.text and not args.batch_file:
        print("ERROR: Must provide either --text or --batch_file", file=sys.stderr)
        sys.exit(1)
    
    # Load model
    model = load_model(args.model)
    
    if args.text:
        # Single text mode
        embedding = embed_single(args.text, model)
        result = {"embedding": embedding, "dim": len(embedding)}
        
        if args.output:
            with open(args.output, "w") as f:
                json.dump(result, f)
        else:
            print(json.dumps(result))
    
    elif args.batch_file:
        # Batch mode
        items = []
        with open(args.batch_file, "r") as f:
            for line in f:
                line = line.strip()
                if not line:
                    continue
                items.append(json.loads(line))
        
        print(f"[Local Embed] Processing {len(items)} items...", file=sys.stderr)
        
        results = embed_batch(items, model)
        
        if args.output:
            with open(args.output, "w") as f:
                for result in results:
                    f.write(json.dumps(result) + "\n")
        else:
            for result in results:
                print(json.dumps(result))
        
        print(f"[Local Embed] ✓ Generated {len(results)} embeddings", file=sys.stderr)


if __name__ == "__main__":
    main()

File: src/lib/learning/vector_index/local_embed.ts
/**
 * Local Embeddings (Node Wrapper)
 * 
 * Wraps Python sentence-transformers script.
 * Generates 384-d embeddings using all-MiniLM-L6-v2.
 */

import { spawn } from "child_process";
import path from "path";

export interface LocalEmbedConfig {
  enabled: boolean;
  model: string;
  dim: number;
}

export function getLocalEmbedConfig(): LocalEmbedConfig {
  return {
    enabled: process.env.LOCAL_EMBEDDINGS_ENABLED === "true",
    model: process.env.LOCAL_EMBEDDINGS_MODEL ?? "sentence-transformers/all-MiniLM-L6-v2",
    dim: parseInt(process.env.LOCAL_EMBEDDINGS_DIM ?? "384", 10),
  };
}

/**
 * Generate Local Embedding
 * 
 * @param text - Text to embed
 * @returns 384-d embedding vector or null if unavailable
 */
export async function generateLocalEmbedding(text: string): Promise<number[] | null> {
  const config = getLocalEmbedConfig();
  
  if (!config.enabled) {
    return null;
  }
  
  if (process.env.REQUIRE_PYTHON_WIRING === "1" && process.env.REQUIRE_PYTHON_WIRING !== "1") {
    console.warn("[Local Embed] Python wiring required but not enabled");
    return null;
  }
  
  const scriptPath = path.resolve(__dirname, "local_embed.py");
  
  return new Promise<number[] | null>((resolve) => {
    const args = [
      scriptPath,
      "--text",
      text,
      "--model",
      config.model,
    ];
    
    const proc = spawn("python", args, {
      stdio: ["ignore", "pipe", "pipe"],
    });
    
    let stdout = "";
    let stderr = "";
    
    proc.stdout?.on("data", (chunk) => {
      stdout += chunk.toString();
    });
    
    proc.stderr?.on("data", (chunk) => {
      stderr += chunk.toString();
    });
    
    proc.on("close", (code) => {
      if (code === 0) {
        try {
          const result = JSON.parse(stdout);
          const embedding = result.embedding;
          
          if (Array.isArray(embedding) && embedding.length === config.dim) {
            resolve(embedding);
          } else {
            console.error(`[Local Embed] Invalid embedding dim: ${embedding?.length} (expected ${config.dim})`);
            resolve(null);
          }
        } catch (err) {
          console.error("[Local Embed] Failed to parse result:", err);
          resolve(null);
        }
      } else {
        console.error(`[Local Embed] Script failed (code ${code}):`, stderr);
        resolve(null);
      }
    });
    
    proc.on("error", (err) => {
      console.error("[Local Embed] Spawn error:", err);
      resolve(null);
    });
  });
}

/**
 * Generate Batch Embeddings
 * 
 * @param items - Array of {id, text} items
 * @returns Array of {id, embedding} or null if unavailable
 */
export async function generateBatchEmbeddings(
  items: { id: string; text: string }[]
): Promise<{ id: string; embedding: number[] }[] | null> {
  const config = getLocalEmbedConfig();
  
  if (!config.enabled) {
    return null;
  }
  
  // Write batch file
  const { promises: fs } = await import("fs");
  const { tmpdir } = await import("os");
  const batchPath = path.join(tmpdir(), `local_embed_batch_${Date.now()}.jsonl`);
  
  const batchLines = items.map((item) => JSON.stringify(item)).join("\n");
  await fs.writeFile(batchPath, batchLines + "\n", "utf-8");
  
  const scriptPath = path.resolve(__dirname, "local_embed.py");
  
  return new Promise<{ id: string; embedding: number[] }[] | null>(async (resolve) => {
    const args = [
      scriptPath,
      "--batch_file",
      batchPath,
      "--model",
      config.model,
    ];
    
    const proc = spawn("python", args, {
      stdio: ["ignore", "pipe", "pipe"],
    });
    
    let stdout = "";
    
    proc.stdout?.on("data", (chunk) => {
      stdout += chunk.toString();
    });
    
    proc.stderr?.on("data", (chunk) => {
      // Ignore stderr (progress bar noise)
      void chunk;
    });
    
    proc.on("close", async (code) => {
      // Clean up batch file
      await fs.unlink(batchPath).catch(() => {});
      
      if (code === 0) {
        try {
          const lines = stdout.split("\n").filter((line) => line.trim());
          const results = lines.map((line) => JSON.parse(line));
          resolve(results);
        } catch (err) {
          console.error("[Local Embed] Failed to parse batch results:", err);
          resolve(null);
        }
      } else {
        console.error(`[Local Embed] Batch failed (code ${code})`);
        resolve(null);
      }
    });
    
    proc.on("error", async (err) => {
      console.error("[Local Embed] Spawn error:", err);
      await fs.unlink(batchPath).catch(() => {});
      resolve(null);
    });
  });
}
